code/planning/src/planning/preplanning_trajectory.py

```diff
@@ -0,0 +1,891 @@
+                for width in widths:
+                    point, v = self.update_one_point(point1_x, point1_y,
+                                                     point2_x, point2_y,
+                                                     point3_x, point3_y,
+                                                     direction, width)
+                    dist = help_functions.euclid_dist(ref_point, point)
+                    # Choose the width value for the correct lane
+                    if dist < min_dist:
+                        min_dist = dist
+                        min_width = width
+                # if we need to change the lane for the next action
+                # we want a smooth change in the trajectory points
+                points = reference_line
+                last_width = widths[size-1]
+                diff = abs(min_width - last_width)
+                step_size = STEP_SIZE
+                i = 0
+                # if diff != 0 we need to change the lane and calculate
+                # a smooth change
+                if diff != 0:
+                    steps = int(diff / step_size)
+                    first_widths = [last_width - step_size * i
+                                    for i in range(steps)]
+                    for i in range(len(first_widths)):
+                        p1_x = points[0][i]
+                        p1_y = points[1][i]
+                        p2_x = points[0][i+1]
+                        p2_y = points[1][i+1]
+                        if i != len((points[0])) - 2:
+                            p3_x = points[0][i + 2]
+                            p3_y = points[1][i + 2]
+                        point, v = self.update_one_point(p1_x, p1_y,
+                                                         p2_x, p2_y,
+                                                         p3_x, p3_y,
+                                                         direction,
+                                                         first_widths[i])
+                        points[0][i] = point[0]
+                        points[1][i] = point[1]
+                    p = [points[0][i+1:], points[1][i+1:], points[2][i+1:],
+                         points[3][i+1:]]
+                    points = self.update_points(p, direction, min_width)
+                del points[0][index+1:]
+                del points[1][index+1:]
+                del points[2][index+1:]
+                del points[3][index+1:]
+                self.add_waypoints(points)
+                break
+            self.add_waypoints(points)
+
+    def rad_to_degree(self, radians):
+        """ Convert radians value to degrees
+            :param radians: heading value in rad
+            :return: deg: degree value
+        """
+        radians = abs(radians)
+        deg = degrees(radians)
+        if deg > 360:
+            deg = 360 - degrees(radians)
+        return deg
+
+    def next_action_id(self, action: float, sec_id: int, points: list):
```

Comment:
the given action from the leaderboard is not a float yaw value containing the new global heading of the agent but a so called [road_option](https://github.com/ll7/paf22/blob/main/doc/03_research/04_requirements/02_informations_from_leaderboard.md#route-format). Therefore it would be nice if your code could handle them instead. Alternatively you could use the orientation of the next given waypoint but then to prevent misunderstanding i would rename the actions here.

Reply:
yes i have to adapt the Road_option in the implementation. Think that is not a big problem

The other point in the first comment:
for the trajectory we only need the x and y coordinates. The z value gives the current height. For the orientation the yaw value is sufficient (the yaw value is the angle which rotates the car arround the z-axis).